controllers/clustertopology_controller_compute.go

```diff
@@ -69,10 +73,116 @@ type machineDeploymentTopologyState struct {
+		return nil, err
+	}
+
+	if class.clusterClass.Spec.ControlPlane.MachineInfrastructure != nil && class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref != nil {
+		// get ClusterClass.spec.controlPlane.machineInfrastructure
+		if class.controlPlane.infrastructureMachineTemplate, err = r.convertReferenceAPIContractAndGet(ctx, class.clusterClass.Spec.ControlPlane.MachineInfrastructure.Ref); err != nil {
+			return nil, err
+		}
+	}
+
+	// if ControlPlaneTemplate.spec.template.spec.machineTemplate.infrastructureRef is set, convert and patch it
+	// returns an error if the ObjectReference is partially set
+	_, ok, err := unstructured.NestedMap(class.controlPlane.template.Object, "spec", "template", "spec", "machineTemplate", "infrastructureRef")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to check existence of spec.template.spec.machineTemplate.infrastructureRef in %s %q in namespace %q", class.controlPlane.template.GetKind(), class.controlPlane.template.GetName(), class.controlPlane.template.GetNamespace())
+	}
+	if ok {
+		controlPlaneInfrastructureMachineTemplateRef, err := getNestedRef(class.controlPlane.template, "spec", "template", "spec", "machineTemplate", "infrastructureRef")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get spec.template.spec.machineTemplate.infrastructureRef from %s %q in namespace %q", class.controlPlane.template.GetKind(), class.controlPlane.template.GetName(), class.controlPlane.template.GetNamespace())
+		}
+
+		// We use the patchHelper to patch potential changes to the ObjectReference in ControlPlane.
+		controlPlaneTemplatePatchHelper, err := patch.NewHelper(class.controlPlane.template, r.UnstructuredCachingClient)
+		if err != nil {
+			return nil, err
+		}
+		if err := utilconversion.ConvertReferenceAPIContract(ctx, r.Client, r.restConfig, controlPlaneInfrastructureMachineTemplateRef); err != nil {
+			return nil, err
+		}
+		ref := map[string]interface{}{
+			"kind":       controlPlaneInfrastructureMachineTemplateRef.Kind,
+			"namespace":  controlPlaneInfrastructureMachineTemplateRef.Namespace,
+			"name":       controlPlaneInfrastructureMachineTemplateRef.Name,
+			"apiVersion": controlPlaneInfrastructureMachineTemplateRef.APIVersion,
+		}
+		err = unstructured.SetNestedField(class.controlPlane.template.Object, ref, "spec", "template", "spec", "machineTemplate", "infrastructureRef")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to set field .spec.template.spec.machineTemplate.infrastructureRef in %s %q in namespace %q", class.controlPlane.template.GetKind(), class.controlPlane.template.GetName(), class.controlPlane.template.GetNamespace())
+		}
+		if err := controlPlaneTemplatePatchHelper.Patch(ctx, class.controlPlane.template); err != nil {
+			return nil, errors.Wrapf(err, "failed to patch %s %q in namespace %q", class.controlPlane.template.GetKind(), class.controlPlane.template.GetName(), class.controlPlane.template.GetNamespace())
+		}
+	}
```

Comment:
Could we split out this code in another function?

The above is very verbose (and hard to read, lots of horizontal lines). Can we simplify it a bit?

Reply:
Yup, but I would like to discuss first if we want to keep it: https://github.com/kubernetes-sigs/cluster-api/pull/5007#discussion_r675452485

Reply:
Dropped the code